internal/gatewayapi/tls.go

```diff
@@ -9,53 +9,169 @@ import (
+				parseErr = fmt.Errorf("unable to parse PKCS1 formatted private key in %s", corev1.TLSPrivateKeyKey)
+			}
+		case "EC PRIVATE KEY":
+			_, err := x509.ParseECPrivateKey(keyBlock.Bytes)
+			if err != nil {
+				parseErr = fmt.Errorf("unable to parse EC formatted private key in %s", corev1.TLSPrivateKeyKey)
+			}
+		default:
+			return fmt.Errorf("%s key format found in %s, supported formats are PKCS1, PKCS8 or EC", keyBlock.Type, corev1.TLSPrivateKeyKey)
+		}
 	}
 
-	_, err := x509.ParseCertificate(certBlock.Bytes)
-	if err != nil {
-		return fmt.Errorf("unable to parse certificate in %s: %w", corev1.TLSCertKey, err)
+	return parseErr
+}
+
+func getFQDN(cert *x509.Certificate) string {
+	fqdn := ""
+	if len(cert.DNSNames) > 0 {
+		for _, name := range cert.DNSNames {
+			fqdn += name + ","
+		}
+	} else {
+		fqdn = cert.Subject.CommonName
 	}
+	return fqdn
+}
 
-	keyData := secret.Data[corev1.TLSPrivateKeyKey]
+// Since Go 1.9 CommonName is deprecated for validation in the x509.Certificate
+// https://golang.google.cn/doc/go1.15#commonname
+// Restoring this behavior to support certificates without SAN.
+func verifyHostname(cert *x509.Certificate, host string) error {
+	if host == "*" {
+		return nil
+	}
 
-	keyBlock, _ := pem.Decode(keyData)
-	if keyBlock == nil {
-		return fmt.Errorf("unable to decode pem data in %s", corev1.TLSPrivateKeyKey)
+	lowered := toLowerCaseASCII(host)
+	if len(cert.DNSNames) > 0 {
+		for _, match := range cert.DNSNames {
+			if matchHostnames(toLowerCaseASCII(match), lowered) {
+				return nil
+			}
+		}
+	} else if matchHostnames(toLowerCaseASCII(cert.Subject.CommonName), lowered) {
+		return nil
 	}
 
-	var parseErr error
+	return x509.HostnameError{Certificate: cert, Host: host}
+}
 
-	switch keyBlock.Type {
-	case "PRIVATE KEY":
-		_, err := x509.ParsePKCS8PrivateKey(keyBlock.Bytes)
-		if err != nil {
-			parseErr = fmt.Errorf("unable to parse PKCS8 formatted private key in %s", corev1.TLSPrivateKeyKey)
+func matchHostnames(pattern, host string) bool {
+	host = strings.TrimSuffix(host, ".")
+	pattern = strings.TrimSuffix(pattern, ".")
+
+	if len(pattern) == 0 || len(host) == 0 {
+		return false
+	}
+
+	patternParts := strings.Split(pattern, ".")
+	hostParts := strings.Split(host, ".")
+
+	if len(patternParts) != len(hostParts) {
+		return false
+	}
+
+	for i, patternPart := range patternParts {
+		if i == 0 && patternPart == "*" {
+			continue
 		}
-	case "RSA PRIVATE KEY":
-		_, err := x509.ParsePKCS1PrivateKey(keyBlock.Bytes)
-		if err != nil {
-			parseErr = fmt.Errorf("unable to parse PKCS1 formatted private key in %s", corev1.TLSPrivateKeyKey)
+		if patternPart != hostParts[i] {
+			return false
 		}
-	case "EC PRIVATE KEY":
-		_, err := x509.ParseECPrivateKey(keyBlock.Bytes)
-		if err != nil {
-			parseErr = fmt.Errorf("unable to parse EC formatted private key in %s", corev1.TLSPrivateKeyKey)
+	}
+
+	return true
+}
+
+// toLowerCaseASCII returns a lower-case version of in. See RFC 6125 6.4.1. We use
```

Comment:
trying to better understand why a custom fun is needed instead of using `strings.ToLower` ?

Reply:
so basically I wanted to use this [function](https://pkg.go.dev/crypto/x509#Certificate.VerifyHostname) to validate hostnames in x509.Certificate.
But since Go 1.9 CommonName is deprecated for validation in the x509.Certificate
https://golang.google.cn/doc/go1.15#commonname

I wanted to restore this behavior to support certificates without SAN and wanted to copy all logic from crypto/x509.
But I would also be okay with simplifying this function only to our needs.

Reply:
ah either works, rewriting to use `strings.Lower` or adding a link to the orig code and updating the file with the copyright

Reply:
I have added the comment already, but will also link orginal code to make it more visible